README.md
[![banner](https://particles.js.org/images/banner3.png)](https://particles.js.org)

# @tsparticles/react

[![npm](https://img.shields.io/npm/v/@tsparticles/react)](https://www.npmjs.com/package/@tsparticles/react) [![npm](https://img.shields.io/npm/dm/@tsparticles/react)](https://www.npmjs.com/package/@tsparticles/react) [![GitHub Sponsors](https://img.shields.io/github/sponsors/matteobruni)](https://github.com/sponsors/matteobruni)

Official [tsParticles](https://github.com/tsparticles/tsparticles) ReactJS component

[![Slack](https://particles.js.org/images/slack.png)](https://join.slack.com/t/tsparticles/shared_invite/enQtOTcxNTQxNjQ4NzkxLWE2MTZhZWExMWRmOWI5MTMxNjczOGE1Yjk0MjViYjdkYTUzODM3OTc5MGQ5MjFlODc4MzE0N2Q1OWQxZDc1YzI) [![Discord](https://particles.js.org/images/discord.png)](https://discord.gg/hACwv45Hme) [![Telegram](https://particles.js.org/images/telegram.png)](https://t.me/tsparticles)

[![tsParticles Product Hunt](https://api.producthunt.com/widgets/embed-image/v1/featured.svg?post_id=186113&theme=light)](https://www.producthunt.com/posts/tsparticles?utm_source=badge-featured&utm_medium=badge&utm_souce=badge-tsparticles") <a href="https://www.buymeacoffee.com/matteobruni"><img src="https://img.buymeacoffee.com/button-api/?text=Buy me a beer&emoji=🍺&slug=matteobruni&button_colour=5F7FFF&font_colour=ffffff&font_family=Arial&outline_colour=000000&coffee_colour=FFDD00"></a>

## Installation

```shell
npm install @tsparticles/react
```

or

```shell
yarn add @tsparticles/react
```

### create-react-app

Starting from version 1.17.0 there are two official `create-react-app` templates:

-   `cra-template-particles`: Simple ReactJS template with full screen particles, using JavaScript
-   `cra-template-particles-typescript`: Simple ReactJS template with full screen particles, using TypeScript

You can simply install them using the `create-react-app` command like this:

```shell
$ create-react-app your_app --template particles
```

or

```shell
$ create-react-app your_app --template particles-typescript
```

## How to use

### Code

Examples:

#### Remote url

##### JavaScript support - url

```jsx
import { useCallback } from "react";
import Particles from "@tsparticles/react";
//import { loadFull } from "tsparticles"; // if you are going to use `loadFull`, install the "tsparticles" package too.
import { loadSlim } from "@tsparticles/slim"; // if you are going to use `loadSlim`, install the "@tsparticles/slim" package too.

const App = () => {
    const particlesInit = useCallback(async engine => {
        console.log(engine);
        // you can initiate the tsParticles instance (engine) here, adding custom shapes or presets
        // this loads the tsparticles package bundle, it's the easiest method for getting everything ready
        // starting from v2 you can add only the features you need reducing the bundle size
        //await loadFull(engine);
        await loadSlim(engine);
    }, []);

    const particlesLoaded = useCallback(async container => {
        await console.log(container);
    }, []);

    return (
        <Particles id="tsparticles" url="http://foo.bar/particles.json" init={particlesInit} loaded={particlesLoaded} />
    );
};
```

##### TypeScript support - url

```typescript jsx
import { useCallback } from "react";
import Particles from "@tsparticles/react";
import type { Container, Engine } from "@tsparticles/engine";
//import { loadFull } from "tsparticles"; // if you are going to use `loadFull`, install the "tsparticles" package too.
import { loadSlim } from "@tsparticles/slim"; // if you are going to use `loadSlim`, install the "@tsparticles/slim" package too.

const App = () => {
    const particlesInit = useCallback(async (engine: Engine) => {
        console.log(engine);

        // you can initialize the tsParticles instance (engine) here, adding custom shapes or presets
        // this loads the tsparticles package bundle, it's the easiest method for getting everything ready
        // starting from v2 you can add only the features you need reducing the bundle size
        //await loadFull(engine);
        await loadSlim(engine);
    }, []);

    const particlesLoaded = useCallback(async (container: Container | undefined) => {
        await console.log(container);
    }, []);

    return (
        <Particles id="tsparticles" url="http://foo.bar/particles.json" init={particlesInit} loaded={particlesLoaded} />
    );
};
```

#### Options object

##### JavaScript support - object

```jsx
import { useCallback } from "react";
import Particles from "@tsparticles/react";
//import { loadFull } from "tsparticles"; // if you are going to use `loadFull`, install the "tsparticles" package too.
import { loadSlim } from "@tsparticles/slim"; // if you are going to use `loadSlim`, install the "@tsparticles/slim" package too.

const App = () => {
    const particlesInit = useCallback(async engine => {
        console.log(engine);
        // you can initiate the tsParticles instance (engine) here, adding custom shapes or presets
        // this loads the tsparticles package bundle, it's the easiest method for getting everything ready
        // starting from v2 you can add only the features you need reducing the bundle size
        //await loadFull(engine);
        await loadSlim(engine);
    }, []);

    const particlesLoaded = useCallback(async container => {
        await console.log(container);
    }, []);

    return (
        <Particles
            id="tsparticles"
            init={particlesInit}
            loaded={particlesLoaded}
            options={{
                background: {
                    color: {
                        value: "#0d47a1",
                    },
                },
                fpsLimit: 120,
                interactivity: {
                    events: {
                        onClick: {
                            enable: true,
                            mode: "push",
                        },
                        onHover: {
                            enable: true,
                            mode: "repulse",
                        },
                        resize: true,
                    },
                    modes: {
                        push: {
                            quantity: 4,
                        },
                        repulse: {
                            distance: 200,
                            duration: 0.4,
                        },
                    },
                },
                particles: {
                    color: {
                        value: "#ffffff",
                    },
                    links: {
                        color: "#ffffff",
                        distance: 150,
                        enable: true,
                        opacity: 0.5,
                        width: 1,
                    },
                    move: {
                        direction: "none",
                        enable: true,
                        outModes: {
                            default: "bounce",
                        },
                        random: false,
                        speed: 6,
                        straight: false,
                    },
                    number: {
                        density: {
                            enable: true,
                            area: 800,
                        },
                        value: 80,
                    },
                    opacity: {
                        value: 0.5,
                    },
                    shape: {
                        type: "circle",
                    },
                    size: {
                        value: { min: 1, max: 5 },
                    },
                },
                detectRetina: true,
            }}
        />
    );
};
```

##### TypeScript support - object

```typescript jsx
import { useCallback } from "react";
import type { Container, Engine } from "@tsparticles/engine";
import Particles from "@tsparticles/react";
//import { loadFull } from "tsparticles"; // if you are going to use `loadFull`, install the "tsparticles" package too.
import { loadSlim } from "@tsparticles/slim"; // if you are going to use `loadSlim`, install the "@tsparticles/slim" package too.

const App = () => {
    const particlesInit = useCallback(async (engine: Engine) => {
        console.log(engine);

        // you can initialize the tsParticles instance (engine) here, adding custom shapes or presets
        // this loads the tsparticles package bundle, it's the easiest method for getting everything ready
        // starting from v2 you can add only the features you need reducing the bundle size
        //await loadFull(engine);
        await loadSlim(engine);
    }, []);

    const particlesLoaded = useCallback(async (container: Container | undefined) => {
        await console.log(container);
    }, []);
    return (
        <Particles
            id="tsparticles"
            init={particlesInit}
            loaded={particlesLoaded}
            options={{
                background: {
                    color: {
                        value: "#0d47a1",
                    },
                },
                fpsLimit: 120,
                interactivity: {
                    events: {
                        onClick: {
                            enable: true,
                            mode: "push",
                        },
                        onHover: {
                            enable: true,
                            mode: "repulse",
                        },
                        resize: true,
                    },
                    modes: {
                        push: {
                            quantity: 4,
                        },
                        repulse: {
                            distance: 200,
                            duration: 0.4,
                        },
                    },
                },
                particles: {
                    color: {
                        value: "#ffffff",
                    },
                    links: {
                        color: "#ffffff",
                        distance: 150,
                        enable: true,
                        opacity: 0.5,
                        width: 1,
                    },
                    move: {
                        direction: "none",
                        enable: true,
                        outModes: {
                            default: "bounce",
                        },
                        random: false,
                        speed: 6,
                        straight: false,
                    },
                    number: {
                        density: {
                            enable: true,
                            area: 800,
                        },
                        value: 80,
                    },
                    opacity: {
                        value: 0.5,
                    },
                    shape: {
                        type: "circle",
                    },
                    size: {
                        value: { min: 1, max: 5 },
                    },
                },
                detectRetina: true,
            }}
        />
    );
};
```

### Props

| Prop            | Type     | Definition                                                                                                                                          |
| --------------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
| id              | string   | The id of the element.                                                                                                                              |
| width           | string   | The width of the canvas.                                                                                                                            |
| height          | string   | The height of the canvas.                                                                                                                           |
| options         | object   | The options of the particles instance.                                                                                                              |
| url             | string   | The remote options url, called using an AJAX request                                                                                                |
| style           | object   | The style of the canvas element.                                                                                                                    |
| className       | string   | The class name of the canvas wrapper.                                                                                                               |
| canvasClassName | string   | the class name of the canvas.                                                                                                                       |
| container       | object   | The instance of the [particles container](https://particles.js.org/docs/classes/tsParticles_Engine.Core_Container.Container.html)                   |
| init            | function | This function is called after the tsParticles instance initialization, the instance is the parameter and you can load custom presets or shapes here |
| loaded          | function | This function is called when particles are correctly loaded in canvas, the current container is the parameter and you can customize it here         |

#### particles.json

Find all configuration options [here](https://particles.js.org/docs/interfaces/tsParticles_Engine.Options_Interfaces_IOptions.IOptions.html).

You can find sample configurations [here](https://github.com/tsparticles/tsparticles/tree/main/utils/configs/src) 📖

## Demos

Preset demos can be found [here](https://particles.js.org/samples/presets/index.html)

There's also a CodePen collection actively maintained and updated [here](https://codepen.io/collection/DPOage)

Report bugs and issues [here](https://github.com/tsparticles/tsparticles/issues)

[tsParticle Website](https://particles.js.org)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [3.0.0-beta.1](https://github.com/tsparticles/react/compare/v2.12.1...v3.0.0-beta.1) (2023-08-11)


### Bug Fixes

* fixed issue with component update ([7043e2c](https://github.com/tsparticles/react/commit/7043e2c5ec615a4c37f558716ece80b366ab696e))





## [2.12.1](https://github.com/tsparticles/react/compare/v3.0.0-beta.1...v2.12.1) (2023-08-04)

**Note:** Version bump only for package @tsparticles/react-workspace





# [3.0.0-beta.1](https://github.com/tsparticles/react/compare/v2.11.0...v3.0.0-beta.1) (2023-08-04)


### Bug Fixes

* **deps:** update dependency @testing-library/react to v14 ([107b5c9](https://github.com/tsparticles/react/commit/107b5c9c76478bb6eb0ae9d1873f62beefc423a9))
* **deps:** update dependency @testing-library/user-event to v14 ([62043e2](https://github.com/tsparticles/react/commit/62043e22da2caefcbccc1dfa563aa01a2f6769b4))
* **deps:** update dependency web-vitals to v3 ([d3b48dc](https://github.com/tsparticles/react/commit/d3b48dcf6927778903314696c3b6b351d4eaed4f))





# [2.11.0](https://github.com/tsparticles/react/compare/v2.10.1...v2.11.0) (2023-07-14)


### Bug Fixes

* fixed container link, closes [#34](https://github.com/tsparticles/react/issues/34) ([a1c206a](https://github.com/tsparticles/react/commit/a1c206af0f0245715b67b6a6eafd906c897acc8e))





## 2.10.1 (2023-06-04)


### Bug Fixes

* changed deep equal library, the previous one had a bug for circular objects and freezed all ([ccb5f12](https://github.com/tsparticles/react/commit/ccb5f124a69ce622fe591fc8944015da64f8f799))
* **deps:** update dependency @next/font to v13.2.1 ([05372e8](https://github.com/tsparticles/react/commit/05372e8b78d10b0e2e1360c046b6eb32c279389f))
* **deps:** update dependency @next/font to v13.2.3 ([abdded8](https://github.com/tsparticles/react/commit/abdded840c5e46cdc428a591ccf6448bd601bd63))
* **deps:** update dependency @next/font to v13.3.0 ([090176d](https://github.com/tsparticles/react/commit/090176de2b16bb7589241bf06fa40f92d0191f4c))
* **deps:** update dependency @next/font to v13.3.1 ([f4c94cb](https://github.com/tsparticles/react/commit/f4c94cb2a72dd32d947273dd8566bc459ec0876d))
* **deps:** update dependency @next/font to v13.3.3 ([d1b0ac6](https://github.com/tsparticles/react/commit/d1b0ac63010c6da4b1ba83ee3982b40e3e509a02))
* **deps:** update dependency @next/font to v13.3.4 ([47aa225](https://github.com/tsparticles/react/commit/47aa225f0e16ef70d6b721fe67472757e0f211da))
* **deps:** update dependency @next/font to v13.4.0 ([62caba7](https://github.com/tsparticles/react/commit/62caba7b5161e7ffd22298cff6efe0318928ee9e))
* **deps:** update dependency @next/font to v13.4.1 ([5b21bdb](https://github.com/tsparticles/react/commit/5b21bdb26fdcd00762cf5043ce6d3f025720462d))


### Features

* updated templates to React 18 ([d9ae354](https://github.com/tsparticles/react/commit/d9ae354d71f245e85009c96007bd59df7bd422c8))


apps/react/src/frames.ts
import react from "../public/assets/react.svg";
import k8s from "../public/assets/k8s.svg";
import code from "../public/assets/code.png";
import smallDeer from "../public/assets/small-deer.svg";
import {
  ClickMode,
  HoverMode,
  type ISourceOptions,
  MoveDirection,
  OutMode,
  ThemeMode,
} from "tsparticles-engine";
import {
  PolygonMaskType,
  PolygonMaskInlineArrangement,
} from "tsparticles-plugin-polygon-mask";

export type TFrame = {
  backgroundColor: string;
  name: string;
  options?: ISourceOptions;
  slug: string;
};

export const frames: TFrame[] = [
  {
    backgroundColor: "linear-gradient(135deg, #606c88 0%, #3f4c6b 100%)",
    name: "Simple",
    options: {
      fpsLimit: 120,
      particles: {
        number: {
          value: 50,
        },
        size: {
          value: 3,
        },
        move: {
          enable: true,
        },
        links: {
          enable: true,
        },
      },
      interactivity: {
        events: {
          onHover: {
            enable: true,
            mode: HoverMode.repulse,
          },
        },
      },
    },
    slug: "simple",
  },
  {
    backgroundColor: "#232741",
    name: "Bubbles",
    options: {
      fpsLimit: 120,
      particles: {
        number: {
          value: 160,
          density: {
            enable: false,
          },
        },
        size: {
          value: { min: 0.3, max: 3 },
          animation: {
            speed: 4,
          },
        },
        links: {
          enable: false,
        },
        move: {
          enable: true,
          random: true,
          speed: 1,
          direction: MoveDirection.top,
          outModes: OutMode.out,
        },
      },
      interactivity: {
        events: {
          onHover: {
            enable: true,
            mode: HoverMode.bubble,
          },
          onClick: {
            enable: true,
            mode: ClickMode.repulse,
          },
        },
        modes: {
          bubble: {
            distance: 250,
            duration: 2,
            size: 0,
            opacity: 0,
          },
          repulse: {
            distance: 400,
            duration: 4,
          },
        },
      },
    },
    slug: "bubbles",
  },
  {
    backgroundColor:
      "linear-gradient(to bottom, #a90329 0%,#8f0222 44%,#6d0019 100%)",
    name: "Snow",
    options: {
      fpsLimit: 120,
      particles: {
        number: {
          value: 160,
          density: {
            enable: false,
          },
        },
        size: {
          value: { min: 1, max: 10 },
        },
        move: {
          enable: true,
          direction: MoveDirection.bottom,
          outModes: OutMode.out,
        },
        links: {
          enable: false,
        },
      },
      interactivity: {
        events: {
          onClick: {
            enable: true,
            mode: ClickMode.remove,
          },
        },
        modes: {
          remove: {
            quantity: 10,
          },
        },
      },
    },
    slug: "snow",
  },
  {
    backgroundColor: "linear-gradient(45deg, #2d364c 0%,#252d3f 100%)",
    name: "Night Sky",
    options: {
      fpsLimit: 120,
      particles: {
        number: {
          value: 60,
          density: {
            enable: true,
            area: 1500,
          },
        },
        links: {
          enable: true,
          opacity: 0.02,
        },
        move: {
          enable: true,
          direction: MoveDirection.right,
          speed: 0.05,
        },
        size: {
          value: 1,
        },
        opacity: {
          value: { min: 0.05, max: 1 },
          animation: {
            enable: true,
            speed: 1,
          },
        },
      },
      interactivity: {
        events: {
          onClick: {
            enable: true,
            mode: ClickMode.push,
          },
        },
        modes: {
          push: {
            quantity: 1,
          },
        },
      },
      detectRetina: true,
    },
    slug: "night-sky",
  },
  {
    backgroundColor: "linear-gradient(45deg, #7d7e7d 0%,#0e0e0e 100%)",
    name: "Multiple images",
    options: {
      fpsLimit: 120,
      particles: {
        number: {
          value: 8,
          density: {
            enable: true,
            area: 800,
          },
        },
        links: {
          enable: false,
        },
        move: {
          enable: true,
          speed: 1,
          outModes: OutMode.out,
        },
        shape: {
          type: ["image", "circle"],
          options: {
            image: [
              {
                src: react,
                height: 20,
                width: 23,
              },
              {
                src: k8s,
                height: 20,
                width: 20,
              },
              {
                src: code,
                height: 20,
                width: 20,
              },
            ],
          },
        },
        color: {
          value: "#CCC",
        },
        size: {
          value: { min: 10, max: 30 },
          animation: {
            enable: true,
            speed: 4,
            sync: false,
          },
        },
      },
      detectRetina: false,
    },
    slug: "images",
  },
  {
    backgroundColor: "linear-gradient(45deg, #45484d 0%,#000000 100%)",
    name: "Polygon mask",
    options: {
      fpsLimit: 60,
      particles: {
        number: {
          value: 200,
          density: {
            enable: false,
          },
        },
        links: {
          enable: true,
          distance: 30,
          opacity: 0.4,
        },
        move: {
          enable: true,
          speed: 1,
          outModes: OutMode.bounce,
        },
        opacity: {
          animation: {
            enable: true,
            speed: 2,
            sync: false,
          },
          value: { min: 0.05, max: 0.4 },
        },
        size: {
          value: 1,
        },
      },
      polygon: {
        enable: true,
        scale: 0.5,
        type: PolygonMaskType.inline,
        move: {
          radius: 10,
        },
        url: smallDeer,
        inline: {
          arrangement: PolygonMaskInlineArrangement.equidistant,
        },
        draw: {
          enable: true,
          stroke: {
            color: "rgba(255, 255, 255, .2)",
          },
        },
      },
      detectRetina: false,
      interactivity: {
        events: {
          onHover: {
            enable: true,
            mode: HoverMode.bubble,
          },
        },
        modes: {
          bubble: {
            size: 6,
            distance: 40,
          },
        },
      },
    },
    slug: "mask",
  },
  {
    backgroundColor: "black",
    name: "Themes",
    options: {
      themes: [
        {
          name: "light",
          default: {
            value: true,
            auto: true,
            mode: ThemeMode.light,
          },
          options: {
            background: {
              color: "#ffffff",
            },
            particles: {
              color: {
                value: ["#000000", "#0000ff"],
              },
            },
          },
        },
        {
          name: "dark",
          default: {
            value: true,
            auto: true,
            mode: ThemeMode.dark,
          },
          options: {
            background: {
              color: "#000000",
            },
            particles: {
              color: {
                value: ["#ffffff", "#ff0000"],
              },
            },
          },
        },
        {
          name: "red",
          options: {
            background: {
              color: "#ff0000",
            },
            particles: {
              color: {
                value: ["#ffffff", "#000000"],
              },
            },
          },
        },
        {
          name: "green",
          options: {
            background: {
              color: "#00ff00",
            },
            particles: {
              color: {
                value: ["#000000", "#0000ff"],
              },
            },
          },
        },
        {
          name: "blue",
          options: {
            background: {
              color: "#0000ff",
            },
            particles: {
              color: {
                value: ["#ffffff", "#00ff00"],
              },
            },
          },
        },
        {
          name: "yellow",
          options: {
            background: {
              color: "#ffff00",
            },
            particles: {
              color: {
                value: ["#000000", "#ff0000"],
              },
            },
          },
        },
        {
          name: "cyan",
          options: {
            background: {
              color: "#00ffff",
            },
            particles: {
              color: {
                value: ["#000000", "#ff00ff"],
              },
            },
          },
        },
        {
          name: "grey",
          options: {
            background: {
              color: "#777777",
            },
            particles: {
              color: {
                value: ["#ffffff", "#000000"],
              },
            },
          },
        },
      ],
      fpsLimit: 60,

      particles: {
        number: {
          value: 30,
          density: {
            enable: true,
            value_area: 800,
          },
        },
        shape: {
          type: ["circle", "square"],
        },
        opacity: {
          value: 1,
        },
        size: {
          value: 30,
          random: {
            enable: true,
            minimumValue: 15,
          },
        },
        rotate: {
          value: 0,
          direction: "random",
          animation: {
            speed: 5,
            enable: true,
          },
        },
        move: {
          enable: true,
          speed: 6,
          direction: "none",
          out_mode: "out",
        },
      },
      interactivity: {
        detectsOn: "canvas",
        events: {
          onHover: {
            enable: true,
            mode: "repulse",
          },
          onClick: {
            enable: true,
            mode: "push",
          },
          resize: true,
        },
      },
      detectRetina: true,
    },
    slug: "themes",
  },
  {
    backgroundColor: "black",
    name: "No particles",
    options: undefined,
    slug: "no-particles",
  },
];
